pyteal/compiler/optimizer/optimizer.py

```diff
@@ -0,0 +1,73 @@
+from typing import List, cast
+
+from pyteal.ir.tealcomponent import TealComponent
+from pyteal.ir.tealop import TealOp
+from pyteal.ir.ops import Op
+
+
+class OptimizeOptions:
+    """An object which specifies optimizations to be performed."""
+
+    use_slot_to_stack: bool
+    iterate_optimizations: bool
+
+    def __init__(
+        self, *, useSlotToStack: bool = False, iterateOptimizations: bool = False
+    ):
+        """Create a new OptimizeOptions object.
+        Args:
+            useSlotToStack (optional): todo
+            iterateOptimizations (optional): todo
+        """
+        self.use_slot_to_stack = useSlotToStack
+        self.iterate_optimizations = iterateOptimizations
+
+
+def has_load_dependencies(teal: List[TealComponent], slot: int, pos: int):
+    for i, op in enumerate(teal):
+        if i == pos:
+            continue
+
+        if type(op) == TealOp and op.op == Op.load and op.args[0] == slot:
+            return True
+
+    return False
+
+
+def apply_slot_to_stack(teal: List[TealComponent]) -> List[TealComponent]:
+    remove = [False] * len(teal)
+    for i, op in enumerate(teal[:-1]):
+        if type(op) != TealOp or op.op != Op.store:
+            continue
+
+        next_op = teal[i + 1]
+        if type(next_op) != TealOp or next_op.op != Op.load:
+            continue
+
+        if op.args[0] != next_op.args[0]:
+            continue
+
+        if not has_load_dependencies(teal, cast(int, op.args[0]), i + 1):
+            remove[i] = remove[i + 1] = True
+
+    return [op for i, op in enumerate(teal) if not remove[i]]
+
+
+def apply_local_optimizations(
+    teal: List[TealComponent], options: OptimizeOptions
+) -> List[TealComponent]:
+    while True:
```

Comment:
@algoidurovic Though it feels unlikely, presumably it's possible optimizations will loop without end.  I'm wondering if you considered setting an upper bound for the number of optimization passes to attempt?

One possibility is to change `iterate_optimizations` from boolean to the maximum pass count.

Reply:
The loop stops iterating when the optimization no longer changes the code, so it's guaranteed to terminate eventually.

Reply:
Leaving a note to document a verbal discussion with @algoidurovic:
* We discussed that a number of possible outcomes may cause infinite looping.  For example, a mistake in optimization algorithm, TEAL programs written in unexpected ways reaching compilation optimization, etc.
* Rather than use a fixed ceiling to bound iterations, we agreed to bound iteration count = program length.  The suggestion avoids a static, value fixed for all programs while also precluding infinite loops.

Reply:
@algoidurovic Thanks for bringing in the feedback via https://github.com/algorand/pyteal/pull/247/commits/a9ba3efd9fea25cab76e96c7d100fe61d120f578.